resolwe/flow/models/utils.py

```diff
@@ -435,32 +435,33 @@ def add_dir_size(obj):
         if data.status in [Data.STATUS_DONE, Data.STATUS_ERROR] and 'size' in obj and not force:
             return
 
-        path = os.path.join(settings.FLOW_EXECUTOR['DATA_DIR'], str(data.pk), obj['dir'])
+        path = os.path.join(settings.FLOW_EXECUTOR['DATA_DIR'], data.get_location_id(), obj['dir'])
         if not os.path.isdir(path):
             raise ValidationError("Referenced dir does not exist ({})".format(path))
 
         obj['size'] = get_dir_size(path)
         obj['total_size'] = obj['size'] + get_refs_size(obj, path)
 
     data_size = 0
-    for field_schema, fields in iterate_fields(data.output, data.process.output_schema):
-        name = field_schema['name']
-        value = fields[name]
-        if 'type' in field_schema:
-            if field_schema['type'].startswith('basic:file:'):
-                add_file_size(value)
-                data_size += value.get('total_size', 0)
-            elif field_schema['type'].startswith('list:basic:file:'):
-                for obj in value:
-                    add_file_size(obj)
-                    data_size += obj.get('total_size', 0)
-            elif field_schema['type'].startswith('basic:dir:'):
-                add_dir_size(value)
-                data_size += value.get('total_size', 0)
-            elif field_schema['type'].startswith('list:basic:dir:'):
-                for obj in value:
-                    add_dir_size(obj)
-                    data_size += obj.get('total_size', 0)
+    if data.pk:
```

Comment:
It's a bugfix and an optimization. Data should be hydrated only if the model instance exists.

Reply:
So, it has to be in a separate commit.

Reply:
I don't have any numbers, but one reason for this change was because tests failed due to location not being created yet.

Reply:
It looks like tests are doing it in a wrong sequence. `output` should not be populated if `DataLocation` doesn't exist. If `output` is empty the for loop should never execute as `iterate_fields` returns an empty list.

Reply:
This method calculates file and directory sizes. If data objects is not yet created, it also means is does not have a location so there's not point in continuing. I changed the if statement to `if data.location:` which makes the if statement self-explanatory.

Reply:
I agree that `data.location` is much better. But file and directory sizes are retrieved from `data.output`. So it would make even more sense to use that - but note that this condition is already (implicitly) included in the following `for`, if `data.output` is empty, the loop will never execute.